File: frontend/src/api/committees.ts
import { apiClient } from "@/api/client";

// ── Types ──────────────────────────────────────────────────────────────────

export interface Committee {
  id: string;
  kind: string;
  name: string;
  description?: string | null;
  is_active: boolean;
  quorum_threshold_pct?: number | null;
  created_at: string;
  updated_at: string;
}

export interface CommitteePage {
  items: Committee[];
  total: number;
  limit: number;
  offset: number;
}

export interface Meeting {
  id: string;
  committee_id: string;
  scheduled_at: string;
  location?: string | null;
  status: string;
  held_at?: string | null;
  protocol_seq?: number | null;
  protocol_year?: number | null;
  members_total?: number | null;
  present_count?: number | null;
  quorum_met?: boolean | null;
  protocol_no?: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface MeetingPage {
  items: Meeting[];
  total: number;
  limit: number;
  offset: number;
}

export interface DecisionTask {
  id: string;
  decision_id: string;
  assignee_person_id?: string | null;
  due_date?: string | null;
  status: string;
  evidence_note?: string | null;
  is_overdue: boolean;
}

export interface Decision {
  id: string;
  text: string;
  decided_at?: string | null;
  [key: string]: unknown;
}

export interface ProtocolDecision {
  decision: Decision;
  tasks: DecisionTask[];
  votes_for?: number;
  votes_against?: number;
  votes_abstain?: number;
  outcome?: "carried" | "rejected" | null;
  votes?: Vote[];
}

export interface Protocol {
  meeting: Meeting;
  decisions: ProtocolDecision[];
}

// ── Срез-2: attendance, votes, protocol journal ────────────────────────────

export interface MemberDetail {
  id: string;
  committee_id: string;
  person_id: string;
  role: string;
  person_fio?: string | null;
}

export interface Attendance {
  id: string;
  meeting_id: string;
  person_id: string;
  present: boolean;
}

export interface Invitation {
  id: string;
  meeting_id: string;
  person_id: string;
  invited_at: string;
}

/**
 * Клиентское зеркало правила кворума (срез-4): порог в процентах (включительно),
 * без порога — строго больше половины.
 */
export const clientQuorum = (
  presentCount: number,
  membersTotal: number,
  thresholdPct?: number | null,
): boolean => {
  if (membersTotal <= 0) return false;
  if (thresholdPct == null) return presentCount * 2 > membersTotal;
  return presentCount * 100 >= thresholdPct * membersTotal;
};

export type VoteChoice = "for" | "against" | "abstain";

export interface Vote {
  id: string;
  decision_id: string;
  person_id: string;
  choice: VoteChoice;
}

export interface DecisionVoteSummary {
  decision_id: string;
  votes_for: number;
  votes_against: number;
  votes_abstain: number;
  outcome: "carried" | "rejected";
  votes: Vote[];
}

export interface ProtocolJournalItem {
  meeting_id: string;
  committee_id: string;
  committee_name: string;
  protocol_no: string;
  held_at: string;
  members_total: number | null;
  present_count: number | null;
  decisions_count: number;
}

export interface ProtocolJournalPage {
  items: ProtocolJournalItem[];
  total: number;
  limit: number;
  offset: number;
}

// ── Срез-3: KPI dashboard ──────────────────────────────────────────────────

export interface CommitteeKpi {
  committees_total: number;
  committees_active: number;
  meetings_planned: number;
  meetings_held: number;
  meetings_cancelled: number;
  decisions_total: number;
  tasks_total: number;
  tasks_open: number;
  tasks_overdue: number;
  tasks_done: number;
  held_meetings: number;
  avg_attendance_pct: number;
  quorum_rate_pct: number;
}

// ── API ────────────────────────────────────────────────────────────────────

const base = "/committees";

export const committeesApi = {
  async list(params: { limit?: number; offset?: number } = {}): Promise<CommitteePage> {
    const r = await apiClient.get<CommitteePage>(base, {
      params: { limit: 100, offset: 0, ...params },
    });
    return r.data;
  },

  async get(id: string): Promise<Committee> {
    return (await apiClient.get<Committee>(`${base}/${id}`)).data;
  },

  async listMeetings(committeeId: string, params: { limit?: number; offset?: number } = {}): Promise<MeetingPage> {
    const r = await apiClient.get<MeetingPage>(`${base}/${committeeId}/meetings`, {
      params: { limit: 100, offset: 0, ...params },
    });
    return r.data;
  },

  async getProtocol(meetingId: string): Promise<Protocol> {
    return (await apiClient.get<Protocol>(`${base}/meetings/${meetingId}/protocol`)).data;
  },

  // ── Write helpers (срез-1 create/update) ──────────────────────────────

  async createCommittee(payload: {
    kind: string;
    name: string;
    description?: string | null;
    is_active?: boolean;
    quorum_threshold_pct?: number | null;
  }): Promise<Committee> {
    return (await apiClient.post<Committee>(base, payload)).data;
  },

  async updateCommittee(
    id: string,
    payload: {
      kind?: string;
      name?: string;
      description?: string | null;
      is_active?: boolean;
      quorum_threshold_pct?: number | null;
    },
  ): Promise<Committee> {
    return (await apiClient.patch<Committee>(`${base}/${id}`, payload)).data;
  },

  async addMember(
    committeeId: string,
    payload: { person_id: string; role?: string },
  ): Promise<{ id: string; committee_id: string; person_id: string; role: string }> {
    return (
      await apiClient.post<{ id: string; committee_id: string; person_id: string; role: string }>(
        `${base}/${committeeId}/members`,
        payload,
      )
    ).data;
  },

  async listMembers(committeeId: string): Promise<MemberDetail[]> {
    return (await apiClient.get<MemberDetail[]>(`${base}/${committeeId}/members`)).data;
  },

  async removeMember(committeeId: string, memberId: string): Promise<void> {
    await apiClient.delete(`${base}/${committeeId}/members/${memberId}`);
  },

  async createMeeting(
    committeeId: string,
    payload: { scheduled_at: string; location?: string | null },
  ): Promise<Meeting> {
    return (await apiClient.post<Meeting>(`${base}/${committeeId}/meetings`, payload)).data;
  },

  async createDecision(
    meetingId: string,
    payload: { text: string; agenda_item_id?: string | null },
  ): Promise<Decision> {
    return (await apiClient.post<Decision>(`${base}/meetings/${meetingId}/decisions`, payload)).data;
  },

  async createTask(
    decisionId: string,
    payload: { assignee_person_id?: string | null; due_date?: string | null; evidence_note?: string | null },
  ): Promise<DecisionTask> {
    return (await apiClient.post<DecisionTask>(`${base}/decisions/${decisionId}/tasks`, payload)).data;
  },

  async updateTask(
    taskId: string,
    payload: {
      assignee_person_id?: string | null;
      due_date?: string | null;
      status?: string;
      evidence_note?: string | null;
    },
  ): Promise<DecisionTask> {
    return (await apiClient.patch<DecisionTask>(`${base}/tasks/${taskId}`, payload)).data;
  },

  // ── Срез-2: attendance, quorum-hold, votes, protocol journal ──────────

  async getAttendance(meetingId: string): Promise<Attendance[]> {
    return (await apiClient.get<Attendance[]>(`${base}/meetings/${meetingId}/attendance`)).data;
  },

  async putAttendance(
    meetingId: string,
    items: { person_id: string; present: boolean }[],
  ): Promise<Attendance[]> {
    return (
      await apiClient.put<Attendance[]>(`${base}/meetings/${meetingId}/attendance`, { items })
    ).data;
  },

  // ── Срез-4: приглашения + печатная форма протокола ────────────────────

  async getInvitations(meetingId: string): Promise<Invitation[]> {
    return (await apiClient.get<Invitation[]>(`${base}/meetings/${meetingId}/invitations`)).data;
  },

  async putInvitations(meetingId: string, personIds: string[]): Promise<Invitation[]> {
    return (
      await apiClient.put<Invitation[]>(`${base}/meetings/${meetingId}/invitations`, {
        person_ids: personIds,
      })
    ).data;
  },

  async downloadProtocolPrint(
    meetingId: string,
    fmt: "docx" | "pdf",
    nameHint?: string,
  ): Promise<void> {
    const { data } = await apiClient.get<Blob>(`${base}/meetings/${meetingId}/protocol/print`, {
      params: { format: fmt },
      responseType: "blob",
    });
    const url = URL.createObjectURL(data);
    const a = document.createElement("a");
    a.href = url;
    a.download = `committee-protocol-${nameHint ?? meetingId}.${fmt}`;
    a.click();
    URL.revokeObjectURL(url);
  },

  async holdMeeting(meetingId: string): Promise<Meeting> {
    return (await apiClient.patch<Meeting>(`${base}/meetings/${meetingId}`, { status: "held" })).data;
  },

  async castVote(decisionId: string, person_id: string, choice: VoteChoice): Promise<Vote> {
    return (
      await apiClient.post<Vote>(`${base}/decisions/${decisionId}/votes`, { person_id, choice })
    ).data;
  },

  async getVotes(decisionId: string): Promise<DecisionVoteSummary> {
    return (await apiClient.get<DecisionVoteSummary>(`${base}/decisions/${decisionId}/votes`)).data;
  },

  async listProtocols(
    params: { committee_id?: string; limit?: number; offset?: number } = {},
  ): Promise<ProtocolJournalPage> {
    const r = await apiClient.get<ProtocolJournalPage>(`${base}/protocols`, {
      params: { limit: 100, offset: 0, ...params },
    });
    return r.data;
  },

  async getKpi(params: { committee_id?: string } = {}): Promise<CommitteeKpi> {
    const r = await apiClient.get<CommitteeKpi>(`${base}/kpi`, { params });
    return r.data;
  },
};

File: backend/app/schemas/committees.py
"""Pydantic schemas for committees срез-1/срез-2 (P10-01)."""

from __future__ import annotations

from datetime import date, datetime

from pydantic import computed_field

from app.models.committees import (
    CommitteeKind,
    CommitteeMemberRole,
    DecisionTaskStatus,
    MeetingStatus,
    VoteChoice,
)
from app.schemas.base import BaseSchema


# --- Committee ---
class CommitteeCreate(BaseSchema):
    kind: CommitteeKind
    name: str
    description: str | None = None
    is_active: bool = True


class CommitteeUpdate(BaseSchema):
    kind: CommitteeKind | None = None
    name: str | None = None
    description: str | None = None
    is_active: bool | None = None


class CommitteeRead(BaseSchema):
    id: str
    kind: CommitteeKind
    name: str
    description: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class CommitteePage(BaseSchema):
    items: list[CommitteeRead]
    total: int
    limit: int
    offset: int


# --- Member ---
class MemberCreate(BaseSchema):
    person_id: str
    role: CommitteeMemberRole = CommitteeMemberRole.MEMBER


class MemberRead(BaseSchema):
    id: str
    committee_id: str
    person_id: str
    role: CommitteeMemberRole


class MemberDetailRead(BaseSchema):
    id: str
    committee_id: str
    person_id: str
    role: CommitteeMemberRole
    person_fio: str | None = None


# --- Meeting ---
class MeetingCreate(BaseSchema):
    scheduled_at: datetime
    location: str | None = None


class MeetingStatusUpdate(BaseSchema):
    status: MeetingStatus


class MeetingRead(BaseSchema):
    id: str
    committee_id: str
    scheduled_at: datetime
    location: str | None
    status: MeetingStatus
    held_at: datetime | None = None
    protocol_seq: int | None = None
    protocol_year: int | None = None
    members_total: int | None = None
    present_count: int | None = None
    quorum_met: bool | None = None
    created_at: datetime
    updated_at: datetime

    @computed_field(return_type=str | None)
    @property
    def protocol_no(self) -> str | None:
        if self.protocol_seq is None or self.protocol_year is None:
            return None
        return f"{self.protocol_seq}/{self.protocol_year}"


class MeetingPage(BaseSchema):
    items: list[MeetingRead]
    total: int
    limit: int
    offset: int


# --- Agenda ---
class AgendaItemCreate(BaseSchema):
    title: str
    seq: int = 1
    presenter_person_id: str | None = None


class AgendaItemRead(BaseSchema):
    id: str
    meeting_id: str
    seq: int
    title: str
    presenter_person_id: str | None


# --- Decision ---
class DecisionCreate(BaseSchema):
    text: str
    agenda_item_id: str | None = None


class DecisionRead(BaseSchema):
    id: str
    meeting_id: str
    agenda_item_id: str | None
    text: str
    decided_at: datetime


# --- Decision task ---
class DecisionTaskCreate(BaseSchema):
    assignee_person_id: str | None = None
    due_date: date | None = None
    evidence_note: str | None = None


class DecisionTaskUpdate(BaseSchema):
    assignee_person_id: str | None = None
    due_date: date | None = None
    status: DecisionTaskStatus | None = None
    evidence_note: str | None = None


class DecisionTaskRead(BaseSchema):
    id: str
    decision_id: str
    assignee_person_id: str | None
    due_date: date | None
    status: DecisionTaskStatus
    evidence_note: str | None
    is_overdue: bool


# --- Attendance ---
class AttendanceItem(BaseSchema):
    person_id: str
    present: bool = True


class AttendanceBulkUpdate(BaseSchema):
    items: list[AttendanceItem]


class AttendanceRead(BaseSchema):
    id: str
    meeting_id: str
    person_id: str
    present: bool


# --- Votes ---
class VoteCreate(BaseSchema):
    person_id: str
    choice: VoteChoice


class VoteRead(BaseSchema):
    id: str
    decision_id: str
    person_id: str
    choice: VoteChoice


class DecisionVoteSummary(BaseSchema):
    decision_id: str
    votes_for: int
    votes_against: int
    votes_abstain: int
    outcome: str | None = None  # "carried" | "rejected" | None (no votes yet)
    votes: list[VoteRead]


# --- Protocol journal ---
class ProtocolJournalItem(BaseSchema):
    meeting_id: str
    committee_id: str
    committee_name: str
    protocol_no: str  # "N/YYYY"
    held_at: datetime
    members_total: int | None
    present_count: int | None
    decisions_count: int


class ProtocolJournalPage(BaseSchema):
    items: list[ProtocolJournalItem]
    total: int
    limit: int
    offset: int


# --- Protocol projection ---
class ProtocolDecision(BaseSchema):
    decision: DecisionRead
    tasks: list[DecisionTaskRead]
    votes_for: int = 0
    votes_against: int = 0
    votes_abstain: int = 0
    outcome: str | None = None  # "carried" | "rejected" | None (no votes yet)
    votes: list[VoteRead] = []


class ProtocolRead(BaseSchema):
    meeting: MeetingRead
    decisions: list[ProtocolDecision]
